### ISSUE TITLE
Why the generator, supervisor and discriminator are trained for an extra two steps when jointly training?

### ISSUE THREAD

**Issue by BaeHann** [benearnthof/TimeGAN#1]
In the line 179 of modules_and_training.py, the comments and codes say that "the generator, supervisor and discriminator are trained for an extra two steps". What's the purpose of that?

**Comment by benearnthof** [comment]
Hi @BaeHann
if i remember correctly, the training for TimeGAN (and a lot of other GANs) may get unstable or collapse if the discriminator gets too good too quickly. The MiniMax objective you're trying to optimize is not a convex function. (You can imagine the geometry of the combined generator and discriminator loss as a saddle.) This is why the Generator is trained for an extra step. This is not a practice I'd recommend, some other form of regularization should be employed, but this is what the authors of the original paper did in their tensorflow implementation.

**Comment by BaeHann** [comment]
Hi! Thank you for your insightful answer. Would you think that the incorporation of WGAN will be practical and helpful in the training of Time-GAN？

**Comment by BaeHann** [comment]
> Hi @BaeHann if i remember correctly, the training for TimeGAN (and a lot of other GANs) may get unstable or collapse if the discriminator gets too good too quickly. The MiniMax objective you're trying to optimize is not a convex function. (You can imagine the geometry of the combined generator and discriminator loss as a saddle.) This is why the Generator is trained for an extra step. This is not a practice I'd recommend, some other form of regularization should be employed, but this is what the authors of the original paper did in their tensorflow implementation.
   
Hi! Thank you for your insightful answer. Would you think that the incorporation of WGAN will be practical and helpful in the training of Time-GAN？

**Comment by benearnthof** [comment]
> > Hi @BaeHann if i remember correctly, the training for TimeGAN (and a lot of other GANs) may get unstable or collapse if the discriminator gets too good too quickly. The MiniMax objective you're trying to optimize is not a convex function. (You can imagine the geometry of the combined generator and discriminator loss as a saddle.) This is why the Generator is trained for an extra step. This is not a practice I'd recommend, some other form of regularization should be employed, but this is what the authors of the original paper did in their tensorflow implementation.
> 
> Hi! Thank you for your insightful answer. Would you think that the incorporation of WGAN will be practical and helpful in the training of Time-GAN？

Yes, WGAN should help with stabilizing training as the distance is smooth and has well behaved gradients everywhere. You can check out this article by Lilian Weng for a lot of insight into stabilizing GAN training and a very nice explanation of WGAN https://lilianweng.github.io/posts/2017-08-20-gan/

**Comment by BaeHann** [comment]
Thank you very much!
＼(^ω^)／
